### ISSUE TITLE
Parsing Error at `\old` term.

### ISSUE THREAD

**Issue by wadoon** [OpenJML/OpenJML#597]
The given Java+JML source is not parsed without errors from OpenJML in commit 6e9c37b9da10da1ef52d93c36221f0f4eaead9c9, but parsable by [KeY](https://key-project.org)  and [ESC@rise4fun](https://rise4fun.com/OpenJMLESC/7jd).

```java
class SITA3{
   /*@ public normal_behaviour
     @  ensures  
     @    a[pos1]  == \old(a[pos2]) &&
     @    a[pos2]  == \old(a[pos1]);
     @*/
    public void  swap(int[] a,int pos1, int pos2) {
	int temp;
        temp = a[pos1]; a[pos1] = a[pos2] ; a[pos2] = temp;
    }
}
```
**Error Messages:**
```
SITA3.java:4: error: The type or expression near here is invalid (or not implemented): ( token \old in JmlParser.term3())
     @    a[pos1]  == \old(a[pos2]) &&
                      ^
SITA3.java:5: error: The type or expression near here is invalid (or not implemented): ( token \old in JmlParser.term3())
     @    a[pos2]  == \old(a[pos1]);
                      ^
SITA3.java:1: error: This type declaration (SITA3) is not matched by a binary class
class SITA3{
^
3
```

Greetings from the KeY-Project.

@mattulbrich

---

**Used Stub:**
```java 
import org.jmlspecs.openjml.Factory;
import org.jmlspecs.openjml.IAPI;
import org.jmlspecs.openjml.JmlTree.JmlCompilationUnit;

public class Main {
	public static void main(String[] args) throws Exception {
		IAPI api = Factory.makeAPI(args);
		JmlCompilationUnit cu = api.parseSingleFile("SITA3.java");
		int i = api.typecheck(cu);
		System.out.println(i);
		System.out.println(api.prettyPrint(cu));
	}
}
```

**Comment by davidcok** [comment]
Problem not reproduced; test case is fine in current development branch and will be for 0.8.25

**Comment by wadoon** [comment]
I investigated further. 

**tl;dr** `API::parseFiles` works fine. `API::parseSingleFile` and `API::parseAndCheck` don't.

1. Indeed `java -jar openjml.jar -check SITA3.java` works fine. 
2. Trying to reproduce this by the openjml API I used this snippet and the openjml 0.8.24 libraries. 
     (Only executing one of the given test methods)

```java
import java.io.File;
import java.util.List;

import org.jmlspecs.openjml.API;
import org.jmlspecs.openjml.API.Finder;
import org.jmlspecs.openjml.Factory;
import org.jmlspecs.openjml.IAPI;
import org.jmlspecs.openjml.JmlTree;
import org.jmlspecs.openjml.JmlTree.JmlCompilationUnit;

public class Main {
	static File file = new File("SITA3.java");

	public static void main(String[] args) throws Exception {
                // Select only one: 
		demoCheckFile();
		//doingParseAndCheck();	
		//doingParseFile();
	}

	private static void doingParseAndCheck() throws Exception {
		System.out.println("Main.doingParseAndCheck()");
		IAPI api = Factory.makeAPI(new String[] {});
		api.parseAndCheck(file);
	}

	private static void doingParseFile() throws Exception {
		System.out.println("Main.doingParseFile()");
		IAPI api = Factory.makeAPI(new String[] {});
		JmlCompilationUnit cu = api.parseSingleFile("SITA3.java");
		int i = api.typecheck(cu);
	}

	private static void demoCheckFile() throws Exception {
		// Taken from DemoCheckFiles.java
		System.out.println("Main.demoCheckFile()");
		try {			
			IAPI m = Factory.makeAPI("-noPurityCheck");
			List<JmlTree.JmlCompilationUnit> asts = m.parseFiles(file);
			int ret = m.typecheck(asts);
			System.out.println("Trees: " + asts.size() + "  Errors: " + ret);
		} catch (Exception e) {
			System.out.println(e);
		}
	}
}
```
3. Leading to following output: 
```
Main.demoCheckFile()
Trees: 1  Errors: 0
```

```
Main.doingParseAndCheck()
SITA3.java:4: error: The type or expression near here is invalid (or not implemented): ( token \old in JmlParser.term3())
     @    a[pos1]  == \old(a[pos2]) &&
                      ^
SITA3.java:5: error: The type or expression near here is invalid (or not implemented): ( token \old in JmlParser.term3())
     @    a[pos2]  == \old(a[pos1]);
                      ^
/home/weigl/workspace/openjml-api-test/jmlspecs.jar(java/lang/Object.jml):81: error: The type or expression near here is invalid (or not implemented): ( token \typeof in JmlParser.term3())
      @     requires obj != null && \typeof(this) == \type(Object);
                                    ^
/home/weigl/workspace/openjml-api-test/jmlspecs.jar(java/lang/Object.jml):94: error: The type or expression near here is invalid (or not implemented): ( token \elemtype in JmlParser.term3())
      @   ensures \elemtype(\typeof(\result)) == \elemtype(\typeof(this));
                  ^
/home/weigl/workspace/openjml-api-test/jmlspecs.jar(java/lang/Object.jml):94: error: The type or expression near here is invalid (or not implemented): ( token \typeof in JmlParser.term3())
      @   ensures \elemtype(\typeof(\result)) == \elemtype(\typeof(this));
                            ^
/home/weigl/workspace/openjml-api-test/jmlspecs.jar(java/lang/Object.jml):94: error: The type or expression near here is invalid (or not implemented): ( token \elemtype in JmlParser.term3())
      @   ensures \elemtype(\typeof(\result)) == \elemtype(\typeof(this));
                                                 ^
/home/weigl/workspace/openjml-api-test/jmlspecs.jar(java/lang/Object.jml):94: error: The type or expression near here is invalid (or not implemented): ( token \typeof in JmlParser.term3())
      @   ensures \elemtype(\typeof(\result)) == \elemtype(\typeof(this));
                                                           ^
/home/weigl/workspace/openjml-api-test/jmlspecs.jar(java/lang/Object.jml):102: error: The type or expression near here is invalid (or not implemented): ( token \typeof in JmlParser.term3())
      @   ensures \typeof(\result) == \typeof(this);  // This is not a strict requirement, although generally expected
                  ^
/home/weigl/workspace/openjml-api-test/jmlspecs.jar(java/lang/Object.jml):102: error: The type or expression near here is invalid (or not implemented): ( token \typeof in JmlParser.term3())
      @   ensures \typeof(\result) == \typeof(this);  // This is not a strict requirement, although generally expected
                                      ^
[...]
```

```
Main.doingParseFile()
SITA3.java:4: error: The type or expression near here is invalid (or not implemented): ( token \old in JmlParser.term3())
     @    a[pos1]  == \old(a[pos2]) &&
                      ^
SITA3.java:5: error: The type or expression near here is invalid (or not implemented): ( token \old in JmlParser.term3())
     @    a[pos2]  == \old(a[pos1]);
```

The API behaves differently for the same task (parsing a java file) depending on the called method.

**Comment by davidcok** [comment]
Reproduced problem and fixed for version 0.8.27. Workaround is to always have some options in the makeAPI call.
